llama-index-integrations/embeddings/llama-index-embeddings-vertex/llama_index/embeddings/vertex/base.py

```diff
@@ -0,0 +1,228 @@
+            texts=[query], embed_mode=self.embed_mode, is_query=True
+        )
+        embeddings = self._model.get_embeddings(texts, **self.additional_kwargs)
+        return embeddings[0].values
+
+    async def _aget_query_embedding(self, query: str) -> Embedding:
+        texts = _get_embedding_request(
+            texts=[query], embed_mode=self.embed_mode, is_query=True
+        )
+        embeddings = await self._model.get_embeddings_async(
+            texts, **self.additional_kwargs
+        )
+        return embeddings[0].values
+
+
+class VertexMultiModalEmbedding(MultiModalEmbedding):
+    embed_dimension: int = Field(description="The vertex output embedding dimension.")
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additional kwargs for the Vertex."
+    )
+
+    _model: MultiModalEmbeddingModel = PrivateAttr()
+    _embed_dimension: int = PrivateAttr()
+
+    def __init__(
+        self,
+        model_name: str = "multimodalembedding",
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[Any] = None,
+        embed_dimension: int = 1408,
+        embed_batch_size: int = DEFAULT_EMBED_BATCH_SIZE,
+        callback_manager: Optional[CallbackManager] = None,
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        init_vertexai(project=project, location=location, credentials=credentials)
+        callback_manager = callback_manager or CallbackManager([])
+        additional_kwargs = additional_kwargs or {}
+
+        super().__init__(
+            embed_dimension=embed_dimension,
+            additional_kwargs=additional_kwargs,
+            model_name=model_name,
+            embed_batch_size=embed_batch_size,
+            callback_manager=callback_manager,
+        )
+        self._model = MultiModalEmbeddingModel.from_pretrained(model_name)
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "VertexMultiModalEmbedding"
+
+    def _get_text_embedding(self, text: str) -> Embedding:
+        return self._model.get_embeddings(
+            contextual_text=text,
+            dimension=self.embed_dimension,
+            **self.additional_kwargs
+        ).text_embedding
+
+    def _get_image_embedding(self, img_file_path: ImageType) -> Embedding:
+        if isinstance(img_file_path, str):
+            image = Image.load_from_file(img_file_path)
+        else:
+            image = Image(image_bytes=img_file_path.getvalue())
+        embeddings = self._model.get_embeddings(
+            image=image, dimension=self.embed_dimension, **self.additional_kwargs
+        )
+        return embeddings.image_embedding
+
+    def _get_query_embedding(self, query: str) -> Embedding:
+        return self._get_text_embedding(query)
+
+    # Vertex AI SDK does not support async variants yet
+    async def _aget_text_embedding(self, text: str) -> Embedding:
+        return self._get_text_embedding(text)
```

Comment:
interesting, the normal text embeddings have async but multi modal doesn't?

Reply:
Unforunately they haven't implemented in the highlevel Vertex AI SDK. It is possible to do async prediction using the client library. 

I can either leave this as it, or subclass `MultiModalEmbeddingModel` and implmenet the async variant?

Reply:
nah no need to sublcass, thats ok for now

Reply:
After looking a bit more, `aiplatform` limits future concurrency internally, and wait for all futures to complete before doing the next async prediction and locks [https://github.com/googleapis/python-aiplatform/blob/63ad1bf9e365d2f10b91e2fd036e3b7d937336c0/google/cloud/aiplatform/models.py#L1641]()